### DISCUSSION TITLE
Network_mode for caddy/reverseproxy container

### DISCUSSION THREAD

**Discussion by loeffelpan** [nextcloud/all-in-one#2043] [Questions] [answered]
I'm running AIO with caddy as reverseproxy.
For now `network_mode` of the caddy-container is `host`.
Is this nessecary?

I tried to change the `Caddyfile` using host's IP instead of localhost. But NC gives me an error because unknown reverse proxy.

I noticed NC is only configured on every start to accept ony 127.0.0.1 as reverse proxy.
But the apache is on another container, so 127.0.0.1 shouldn't work from my point of view.

Can you please help with configuration of `APACHE_IP_BINDNG` and Caddyfile when caddy runs in `bridge` mode?

Thanks in advice.

**Comment by Zoey2936** [comment, marked as answer]
The short answer: use host mode with IP binding to localhost

The long: caddy needs to connect to an IP and your server mostly only provides localhost and one external IP, using the external IP in bridge mode would be possible, but is insecure, so the best option is localhost, but since the Caddy container has different localhost then the host in bridge, it can on connect to localhost, if Caddy runs in bridge mode. The only other option would be, if you are behind a NAT, you could connect caddy to the private IP of your NAT and bind Nextcloud to it, so the Nextcloud is only exposed to your LAN and only Caddy is accessible from the internet and not the APACHE_PORT

**Comment by loeffelpan** [comment]
I would like to describe my (network) setup to understand what the best / correct solution.

Caddy and Nextcloud are both on docker in my local network behind my router on the same ubuntu server. Router does naturally NAT and only Caddy is accessible from the internet with port 80 and 443.
Caddyfile in host mode has two reverseproxy sites for nextcloud to localhost:11000 and for aio-interface to localhost:8443.

If Caddy goes to bridge mode a have to change to host.ip instead of localhost, as localhost inside caddy's container wouldn't reach nextcloud or aio. But after that I got those unknown reverse proxy errors in nextcloud. No matter what I choose for the IP binding.

I don't get how to tell nextcloud that another reverse proxy than 127.0.0.1 is allowed even, when I use another IP binding for apache.

Maybe you can help with these information.

**Reply by Zoey2936** [reply]
Did you tried to set apache ip binding to 0.0.0.0?

**Reply by loeffelpan** [reply]
I tried again with no `APACHE_IP_BINDING` and caddy in bridge mode. No error messages in AIO container logs. But Caddy gets `connection refused` from `172.20.2.2:11000` which is the apache.

If host mode is the best option, I will keep it like that. But most recommendations say to run any container in bridge mode to use dockers isolation benefits.

**Reply by Zoey2936** [reply]
and with ip binding to 0.0.0.0?

**Reply by loeffelpan** [reply]
Unfortunately the same.
But should work with this configuration, or am I wrong?

**Reply by Zoey2936** [reply]
Can you please try again with ip binding set to 0.0.0.0 and send me the output of this: `docker inspect nextcloud-aio-apache`?

**Reply by loeffelpan** [reply]
I don't know why, but suddenly nextcloud (apache) is working as expected with caddy in bridge mode an apache with binding on 0.0.0.0.
But AIO is not reachable by now. That one goes through caddy as reverse proxy to `https://172.20.2.2:8080`. No errors in the log.

**Reply by Zoey2936** [reply]
can you run `curl -LI nated-private-ip:11000`?

**Reply by loeffelpan** [reply]
Sorry, edited my last response.
Opening `https://172.20.2.2:8080/` in browser is working.

**Reply by Zoey2936** [reply]
so it works now?

**Reply by loeffelpan** [reply]
Nextcloud via caddy and apache-container works with caddy in bridge mode - suddenly. No idea what I've done before.

But AIO-interface does not work. Same domain on Port 8443 (as recommended in the docs) with same caddy as revese proxy.
```
https://kbw3.duckdns.org:8443 {
  reverse_proxy https://172.20.2.2:8080 {
    transport http {
      tls_insecure_skip_verify
    }
  }
}
```
Browser just gets timeout. No error in caddy logs. Direct access on `https://172.20.2.2:8080` works.

**Reply by loeffelpan** [reply]
Maybe this is helpful. Done from hosts bash. Seems that caddy in bridge mode can not connect to AIO interface webserver.
```
$ sudo docker exec -it caddy nc -vz 172.20.2.2 8080
172.20.2.2 (172.20.2.2:8080) open  
$ nc -vz 172.20.2.2 8080
Connection to 172.20.2.2 8080 port [tcp/http-alt] succeeded!
```

**Reply by Zoey2936** [reply]
can you try to connect to port 8080 from inside caddy using wget/curl?

**Reply by loeffelpan** [reply]
```
$ wget 127.0.0.1:8080
--2023-02-24 18:56:28--  http://127.0.0.1:8080/
Connecting to 127.0.0.1:8080... connected.
HTTP request sent, awaiting response... 400 Bad Request
2023-02-24 18:56:28 ERROR 400: Bad Request.

$ wget 172.20.2.2:8080
--2023-02-24 18:56:54--  http://172.20.2.2:8080/
Connecting to 172.20.2.2:8080... connected.
HTTP request sent, awaiting response... 400 Bad Request
2023-02-24 18:56:54 ERROR 400: Bad Request.

$ sudo docker exec -it caddy wget 172.20.2.2:8080
Connecting to 172.20.2.2:8080 (172.20.2.2:8080)
wget: server returned error: HTTP/1.1 400 Bad Request
```

Strange. All the same. But `nc` gives different results.

**Reply by Zoey2936** [reply]
can you try with https:// ?

**Reply by loeffelpan** [reply]
Slightly different but both good.
```
$ sudo docker exec -it caddy wget --no-check-certificate https://172.20.2.2:8080
Connecting to 172.20.2.2:8080 (172.20.2.2:8080)
saving to 'index.html'
'index.html' saved

$ wget --no-check-certificate https://172.20.2.2:8080
--2023-02-24 19:57:19--  https://172.20.2.2:8080/
Connecting to 172.20.2.2:8080... connected.
WARNING: cannot verify 172.20.2.2's certificate, issued by ‘CN=nextcloud.local,O=Dev,L=Local,ST=BE,C=DE’:
  Self-signed certificate encountered.
    WARNING: certificate common name ‘nextcloud.local’ doesn't match requested host name ‘172.20.2.2’.
HTTP request sent, awaiting response... 302 Found
Location: /login [following]
--2023-02-24 19:57:19--  https://172.20.2.2:8080/login
Reusing existing connection to 172.20.2.2:8080.
HTTP request sent, awaiting response... 200 OK
Saving to: ‘index.html’
2023-02-24 19:57:19 (38,5 MB/s) - ‘index.html’ saved [1316]
```

**Reply by Zoey2936** [reply]
can you show your caddy config again?

**Reply by loeffelpan** [reply]
Sure.
```
https://kbw3.duckdns.org {
  rewrite /.well-known/carddav /remote.php/dav
  rewrite /.well-known/caldav /remote.php/dav
  reverse_proxy 172.20.2.2:11000
}

https://kbw3.duckdns.org:8443 {
  reverse_proxy https://172.20.2.2:8080 {
    transport http {
      tls_insecure_skip_verify
    }
  }
}
```

**Reply by Zoey2936** [reply]
is port 8443 open in your firewall?

**Reply by loeffelpan** [reply]
Sure. Without that caddy in host mode won't work, too. But you gave me a lead and I figured it out.
I forgot the portmapping for 8443 in the caddy container when switching to bridge mode.
Sorry for that quite long troubleshooting.

Summary:
To use caddy reverse proxy in `bridge` mode I had to set `APACHE_IP_BINDING` to 0.0.0.0 or omit this.
Bind caddy container to 80, 443 and 8443. Thats all.

**Reply by Zoey2936** [reply]
so it works?

**Reply by loeffelpan** [reply]
Yes it does. Thanks for your help.
